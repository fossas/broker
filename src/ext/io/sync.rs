//! Types and functions for IO actions, wrapped in Broker errors and semantics.
//!
//! # Async
//!
//! Generally, prefer the variants in the parent module, as they are async.
//!
//! Functions in this module are the backing sync variants of those functions
//! and should only be used if inside a sync context.
//!
//! # Why backing sync
//!
//! Rust standard library IO operations are synchronous,
//! so in order to make them "async" we have to run these synchronous operations
//! in Tokio's backing thread pool.
//!
//! # Duplication
//!
//! At present, this sync module and its async parent are manually written
//! and contain a fair amount of duplication. Over time I'd like to unify them
//! using a macro to autogenerate the async versions,
//! or swapping to a backing FS implementation that is async native.

use std::{
    env, fmt, fs, iter,
    path::{Path, PathBuf},
};

use error_stack::{IntoReport, Report, ResultExt};
use once_cell::sync::OnceCell;
use tracing::debug;

use crate::ext::{
    error_stack::{DescribeContext, ErrorHelper, IntoContext},
    iter::{AlternativeIter, ChainOnceWithIter},
    result::{WrapErr, WrapOk},
};

/// Errors that are possibly surfaced during IO actions.
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// The provided path-like item failed validation.
    /// Often these errors are related to permissions or the path not existing.
    #[error("validate path")]
    ValidatePath,

    /// The provided file path does not reference a file on disk.
    #[error("path is not a regular file")]
    NotRegularFile,

    /// Failed to locate the HOME directory for the current user.
    #[error("locate home directory for the current user")]
    LocateUserHome,

    /// Failed to locate the current working directory.
    #[error("locate working directory")]
    LocateWorkingDirectory,

    /// Failed to read the contents of the file at the provided path.
    #[error("read contents of file")]
    ReadFileContent,

    /// The value was already set.
    #[error("value already set: {0}")]
    ValueAlreadySet(String),
}

static DATA_ROOT: OnceCell<PathBuf> = OnceCell::new();

/// The root data directory for Broker.
/// Broker uses this directory to store working state and to read configuration information.
///
/// - On Linux and macOS: `~/.config/fossa/broker/`
/// - On Windows: `%USERPROFILE%\.config\fossa\broker`
///
/// Users may also customize this root via [`set_data_root`].
#[tracing::instrument]
pub fn data_root() -> Result<&'static PathBuf, Report<Error>> {
    DATA_ROOT.get_or_try_init(|| {
        debug!("discovering data root from user home");
        home_dir().map(|home| home.join(".config").join("fossa").join("broker"))
    })
}

/// Configure the data root. See [`data_root`] for more info.
#[tracing::instrument]
pub fn set_data_root<P: Into<PathBuf> + std::fmt::Debug>(root: P) -> Result<(), Report<Error>> {
    DATA_ROOT
        .set(root.into())
        .map_err(|err| Error::ValueAlreadySet(err.to_string_lossy().to_string()))
        .map_err(Report::new)
}

/// Searches configured locations for the file with the provided name.
///
/// Locations searched:
/// - The [`data_root`] location.
/// - The [`working_dir`] location.
#[tracing::instrument]
pub fn find<S: AsRef<str> + fmt::Debug>(name: S) -> Result<PathBuf, Report<Error>> {
    let name = PathBuf::from(name.as_ref());
    iter::once_with(|| working_dir().map(|d| d.join(&name)).and_then(validate_file))
        .chain_once_with(|| data_root().map(|d| d.join(&name)).and_then(validate_file))
        .alternative_fold()
        .describe("searches the working directory and '{USER_DIR}/.config/fossa/broker'")
}

/// Searches configured locations (see [`find`])
/// for one of several provided names, returning the first one that was found.
#[tracing::instrument]
pub fn find_some<V, S>(names: V) -> Result<PathBuf, Report<Error>>
where
    S: AsRef<str> + fmt::Debug,
    V: IntoIterator<Item = S> + fmt::Debug,
{
    names.into_iter().map(find).alternative_fold()
}

/// Reads the provided file content to a string.
#[tracing::instrument]
pub fn read_to_string<P: AsRef<Path> + fmt::Debug>(file: P) -> Result<String, Report<Error>> {
    let file = file.as_ref().to_path_buf();
    fs::read_to_string(file)
        .context(Error::ReadFileContent)
        .help("validate that you have access to the file and that it exists")
}

/// Validate that a file path exists and is a regular file.
#[tracing::instrument]
pub fn validate_file(path: PathBuf) -> Result<PathBuf, Report<Error>> {
    let meta = fs::metadata(&path)
        .context(Error::ValidatePath)
        .describe_lazy(|| format!("validate file: {}", path.display()))
        .help("validate that you have access to the file and that it exists")?;

    if meta.is_file() {
        path.wrap_ok()
    } else {
        Error::NotRegularFile
            .wrap_err()
            .into_report()
            .attach_printable_lazy(|| format!("validate file: {}", path.display()))
    }
}

/// Look up the current working directory.
///
/// This function is lazy and memoized:
/// the lookup is performed the first time on demand
/// and (assuming no error was encountered)
/// that result is saved for future invocations.
#[tracing::instrument]
pub fn working_dir() -> Result<&'static PathBuf, Report<Error>> {
    static LAZY: OnceCell<PathBuf> = OnceCell::new();
    LAZY.get_or_try_init(|| {
        debug!("Performing uncached lookup of working directory");
        env::current_dir()
            .context(Error::LocateWorkingDirectory)
            .describe("on macOS and Linux, this uses the system call 'getcwd'")
            .describe("on Windows, this uses the Windows API call 'GetCurrentDirectoryW'")
            .describe("this kind of error is typically caused by the current user not having access to the working directory")
    })
}

/// Look up the user's home directory.
///
/// This function is lazy and memoized:
/// the lookup is performed the first time on demand
/// and (assuming no error was encountered)
/// that result is saved for future invocations.
#[tracing::instrument]
pub fn home_dir() -> Result<&'static PathBuf, Report<Error>> {
    static LAZY: OnceCell<PathBuf> = OnceCell::new();
    LAZY.get_or_try_init(|| {
        debug!("Performing uncached lookup of home directory");
        dirs::home_dir().ok_or(Error::LocateUserHome).into_report()
            .describe("on macOS and Linux, this uses the $HOME environment variable or the system call 'getpwuid_r'")
            .describe("on Windows, this uses the Windows API call 'SHGetKnownFolderPath'")
            .describe("this is a very rare condition, and it's not likely that Broker will be able to resolve this issue")
    })
}
