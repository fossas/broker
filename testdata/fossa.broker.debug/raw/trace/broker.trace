{"timestamp":"2023-04-05T16:30:24.086526Z","level":"INFO","fields":{"message":"Debug artifacts being stored in '/Users/me/.config/fossa/broker/debug/trace'"},"target":"broker::debug"}
