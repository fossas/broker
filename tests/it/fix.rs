use std::sync::RwLock;

use crate::{
    guard_integration_test,
    helper::{assert_equal_contents, copy_recursive, expand_debug_bundle},
    load_config, set_snapshot_vars, temp_config,
};
use broker::{
    cmd::fix::Logger,
    debug::{bundler::TarGz, Bundle, BundleExport},
};
use insta::assert_snapshot;

/// A logger that prints to stdout and also keeps track of what has been logged so that we can test it
#[derive(Default)]
struct TestLogger {
    output: RwLock<Vec<String>>,
}

impl TestLogger {
    fn output(&self) -> String {
        self.output
            .read()
            .expect("read lock must not be poisoned")
            .join("\n")
    }

    fn new() -> Self {
        Default::default()
    }
}

impl Logger for TestLogger {
    fn log<S: AsRef<str>>(&self, content: S) {
        self.output
            .write()
            .expect("write lock must not be poisoned")
            .push(content.as_ref().to_string());
    }
}

/// git gives slightly different output in CI and locally. These filters hide that difference.
fn fix_output_filters() -> Vec<(&'static str, &'static str)> {
    vec![
        (
            "fatal: could not read Username for 'https://github.com': terminal prompts disabled",
            "{git authentication or missing repo error}",
        ),
        (
            r"remote: Repository not found.\s*fatal: repository '[^']*' not found",
            "{git authentication or missing repo error}",
        ),
    ]
}

#[tokio::test]
async fn with_successful_http_no_auth_integration() {
    guard_integration_test!();

    let (_tmp, _, ctx) = temp_config!(load);

    set_snapshot_vars!();
    let (_, conf) = load_config!(
        "testdata/config/basic-http-no-auth.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let logger = TestLogger::new();
    broker::cmd::fix::main(&ctx, &conf, &logger, BundleExport::Disable)
        .await
        .expect("should run fix");

    insta::with_settings!({ filters => fix_output_filters() },
       {
        assert_snapshot!(logger.output());
       }
    );
}

#[tokio::test]
async fn with_failing_http_no_auth_integration_scan() {
    guard_integration_test!();

    let (_tmp, _, ctx) = temp_config!(load);

    set_snapshot_vars!();
    let (_, conf) = load_config!(
        "testdata/config/basic-http-no-auth-empty-repo.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let logger = TestLogger::new();
    broker::cmd::fix::main(&ctx, &conf, &logger, BundleExport::Disable)
        .await
        .expect("should run fix");

    insta::with_settings!({ filters => fix_output_filters() },
       {
        assert_snapshot!(logger.output());
       }
    );
}

#[tokio::test]
async fn with_failing_http_no_auth_download_cli() {
    guard_integration_test!();

    let (_, _, ctx) = temp_config!(load);

    set_snapshot_vars!();
    let (_, conf) = load_config!(
        "testdata/config/basic-http-no-auth-empty-repo.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let logger = TestLogger::new();
    broker::cmd::fix::main(&ctx, &conf, &logger, BundleExport::Disable)
        .await
        .expect("should run fix");

    insta::with_settings!({ filters => fix_output_filters() },
       {
        assert_snapshot!(logger.output());
       }
    );
}

#[tokio::test]
async fn with_failing_http_basic_auth_integration() {
    guard_integration_test!();

    let (_, _, ctx) = temp_config!(load);

    set_snapshot_vars!();
    let (_, conf) = load_config!(
        "testdata/config/basic-http-basic-bad-repo-name.yml",
        "testdata/database/empty.sqlite"
    )
    .await;
    let logger = TestLogger::new();
    broker::cmd::fix::main(&ctx, &conf, &logger, BundleExport::Disable)
        .await
        .expect("should run fix");

    insta::with_settings!({ filters => fix_output_filters() },
       {
        assert_snapshot!(logger.output());
       }
    );
}

#[tokio::test]
async fn with_failing_http_no_auth_integration() {
    guard_integration_test!();

    let (_, _, ctx) = temp_config!(load);

    set_snapshot_vars!();
    let (_, conf) = load_config!(
        "testdata/config/private-repo-http-no-auth.yml",
        "testdata/database/empty.sqlite"
    )
    .await;

    let logger = TestLogger::new();
    broker::cmd::fix::main(&ctx, &conf, &logger, BundleExport::Disable)
        .await
        .expect("should run fix");

    insta::with_settings!({ filters => fix_output_filters() },
       {
        assert_snapshot!(logger.output());
       }
    );
}

#[tokio::test]
async fn generates_debug_bundle() {
    guard_integration_test!();

    let (tmp, conf, _ctx) = temp_config!(load);
    copy_recursive(
        "testdata/fossa.broker.debug/raw",
        conf.debug().location().as_path(),
    );

    let bundle_target = tmp.path().join("fossa.broker.debug.tar.gz");
    let bundler = TarGz::new().expect("must create bundler");
    let bundle =
        Bundle::collect(conf.debug(), bundler, bundle_target).expect("must collect debug bundle");

    let unpacked = expand_debug_bundle(bundle.location());
    assert_equal_contents("testdata/fossa.broker.debug/bundled", unpacked.path());
}
