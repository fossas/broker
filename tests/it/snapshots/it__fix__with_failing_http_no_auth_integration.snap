---
source: tests/it/fix.rs
expression: logger.output()
---

Diagnosing connection to FOSSA

✅ check fossa API connection with no auth required
❌ check fossa API connection with auth required

Diagnosing connections to configured repositories

❌ http://github.com/github/doesnotexist.git

Errors found while checking integrations

❌ http://github.com/github/doesnotexist.git

Broker encountered an error while trying to connect to your git remote at 'http://github.com/github/doesnotexist.git'.

Broker was unable to connect to this repository. Please make sure that the authentication info and the remote are set correctly in your config.yml file.

You are using http transport with no authentication for this integration. To debug this, please make sure that the following command works:

GIT_TERMINAL_PROMPT='0' GCM_INTERACTIVE='never' GIT_ASKPASS='' git -c credential.helper= ls-remote --quiet http://github.com/github/doesnotexist.git

Full error message from git:

run command: git
args: ["-c", "credential.helper=", "ls-remote", "--quiet", "http://github.com/github/doesnotexist.git"]
env: ["GIT_TERMINAL_PROMPT='0'", "GCM_INTERACTIVE='never'", "GIT_ASKPASS=<REMOVED>"]
status: 128
stdout: ''
stderr: '{git authentication or missing repo error}'



Errors found while checking connection to FOSSA

❌ Error checking connection to FOSSA: GET to fossa endpoint 'https://app.fossa.com/api/cli/organization' with authentication required

Broker received an "Unauthorized" status response from FOSSA. This can mean that the fossa_integration_key configured in your config.yml file is not correct. You can obtain a FOSSA API key by going to Settings => Integrations => API in the FOSSA application.

The URL Broker attempted to connect to was 'https://app.fossa.com/api/cli/organization'. Please make sure you can make a request to that URL. For example, try this curl command:

curl -H "Authorization: Bearer <your fossa api key>" https://app.fossa.com/api/cli/organization

Full error message: HTTP status client error (401 Unauthorized) for url (https://app.fossa.com/api/cli/organization)


Collecting debug bundle

❌ Debug bundle collection disabled.
